stan/math/prim/scal/meta/scalar_type.hpp

```diff
@@ -2,25 +2,50 @@
 #define STAN_MATH_PRIM_SCAL_META_SCALAR_TYPE_HPP
 
 #include <stan/math/prim/scal/meta/value_type.hpp>
+#include <type_traits>
 
 namespace stan {
+
+/**
+ * End of recursion to determine base scalar type of a type
+ * The underlying base scalar type. If T is not a container then this
+ * has a static member nameed type with the type T.
+ * @tparam T the type.
+ */
+template <typename T, typename = void>
+struct scalar_type_base {
+  using type = T;
+};
+
+/**
+ * Recursivly goes through array attributes until base type is found
+ * @tparam T the type to clean and return.
+ */
+template <typename T>
+struct scalar_type_base<
+    T, std::enable_if_t<std::is_const<T>::value || std::is_pointer<T>::value>> {
+  using type = typename scalar_type_base<
+      std::remove_pointer_t<std::remove_cv_t<T>>>::type;
+};
```

Comment:
@bob-carpenter @syclik this is probs the only serious change to point out. I'm assuming the intent of `scalar_type` is to get the underlying value so this also cycles through pointers. Does this seem like it's right?